api/capi/src/nntrainer.cpp

```diff
@@ -483,6 +458,157 @@ int ml_train_optimizer_set_property(ml_train_optimizer_h optimizer, ...) {
   return status;
 }
 
+int ml_train_model_set_dataset(ml_train_model_h model,
+                               ml_train_dataset_h dataset) {
+  int status = ML_ERROR_NONE;
+  ml_train_model *nnmodel;
+  ml_train_dataset *nndataset;
+
+  ML_NNTRAINER_CHECK_MODEL_VALIDATION(nnmodel, model);
+  ML_NNTRAINER_CHECK_DATASET_VALIDATION(nndataset, dataset);
+
+  std::shared_ptr<nntrainer::NeuralNetwork> NN;
+  std::shared_ptr<nntrainer::DataBuffer> data;
+
+  NN = nnmodel->network;
+  data = nndataset->data_buffer;
+
+  returnable f = [&]() { return NN->setDataBuffer(data); };
+
+  status = nntrainer_exception_boundary(f);
+  if (status == ML_ERROR_NONE) {
+    nndataset->in_use = true;
+    if (nnmodel->dataset)
+      nnmodel->dataset->in_use = false;
+    nnmodel->dataset = nndataset;
+  }
+
+  return status;
+}
+
+int ml_train_dataset_create_with_generator(ml_train_dataset_h *dataset,
+                                           ml_train_datagen_cb train_cb,
+                                           ml_train_datagen_cb valid_cb,
+                                           ml_train_datagen_cb test_cb) {
+  int status = ML_ERROR_NONE;
+
+  ml_train_dataset *nndataset = new ml_train_dataset;
+  nndataset->magic = ML_NNTRAINER_MAGIC;
+  nndataset->data_buffer =
+    std::make_shared<nntrainer::DataBufferFromCallback>();
+  nndataset->in_use = false;
+
+  std::shared_ptr<nntrainer::DataBufferFromCallback> callback_buffer =
+    std::static_pointer_cast<nntrainer::DataBufferFromCallback>(
+      nndataset->data_buffer);
+
+  status = callback_buffer->setFunc(nntrainer::BUF_TRAIN, train_cb);
+  if (status != ML_ERROR_NONE) {
+    delete nndataset;
+    return status;
+  }
+
+  status = callback_buffer->setFunc(nntrainer::BUF_VAL, valid_cb);
+  if (status != ML_ERROR_NONE) {
+    delete nndataset;
+    return status;
+  }
+
+  status = callback_buffer->setFunc(nntrainer::BUF_TEST, test_cb);
+  if (status != ML_ERROR_NONE) {
+    delete nndataset;
+    return status;
+  }
+
+  *dataset = nndataset;
+  return status;
+}
+
+int ml_train_dataset_create_with_file(ml_train_dataset_h *dataset,
+                                      const char *train_file,
+                                      const char *valid_file,
+                                      const char *test_file) {
+  int status = ML_ERROR_NONE;
+
+  ml_train_dataset *nndataset = new ml_train_dataset;
+  nndataset->magic = ML_NNTRAINER_MAGIC;
+  nndataset->data_buffer =
+    std::make_shared<nntrainer::DataBufferFromDataFile>();
+  nndataset->in_use = false;
+
+  std::shared_ptr<nntrainer::DataBufferFromDataFile> callback_buffer =
+    std::static_pointer_cast<nntrainer::DataBufferFromDataFile>(
+      nndataset->data_buffer);
+
+  if (train_file) {
```

Comment:
train file can't be null w.r.t documentation, so return error when !train_file or change docs?

Reply:
added returning error.